File: py_payment_api/app/services/forwarder/abstract/single_forwarder.py
from typing import Dict, Any, final
from abc import ABC, abstractmethod
from .forwarder import Forwarder
from ....services import Services, PurchaseStatus, get_services

class SingleForwarder(Forwarder):

    def __init__(self, status: PurchaseStatus) -> None:
        self._service :Services[Any, Any] = get_services()
        self.__status = status

 
    @final
    def forward_event(self, event_data: Dict[str, Any]) -> None:
        """
        Forward the event data.
        """
        order_id = str(event_data.get('order_id', ''))
        if self._service:
            self._service.change_order_status(order_id, self.__status)
        self._on_forward_event(event_data)

    @abstractmethod
    def _on_forward_event(self, event_data: Dict[str, Any]) -> None:
        pass

File: tsPaymentApi/src/services/forwarder/abstract/SingleForwarder.ts
import { Forwarder } from './Forwarder';
import { PurchaseStatus } from '../../../enums';
import { getServices } from '../../index';
import { Services } from '../../Services';

export abstract class SingleForwarder implements Forwarder {
    protected readonly service: Services;
    private readonly status: PurchaseStatus;

    constructor(status: PurchaseStatus) {
        this.service = getServices();
        this.status = status;
    }

    public async forwardEvent(eventData: Record<string, any>): Promise<void> {
        const orderId = String(eventData.order_id || '');
        
        if (this.service) {
            await this.service.changeOrderStatus(orderId, this.status);
        }
        
        await this.onForwardEvent(eventData);
    }

    protected abstract onForwardEvent(eventData: Record<string, any>): Promise<void>;
} 